From 7849aa1e485d27352db63be0506fed8729b16000 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Wed, 4 Jan 2023 09:33:35 +0000
Subject: [PATCH 0883/1646] [Huawei] userswap: introduce
 UFFDIO_COPY_MODE_DIRECT_MAP to map without copying

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

Add a new UFFDIO_COPY mode UFFDIO_COPY_MODE_DIRECT_MAP to map physical
pages without copy_from_user().
We introduce uswap_unmap_anon_page() to unmap an anonymous page and
uswap_map_anon_page() to map page to src addr. We also introduce
mfill_atomic_pte_nocopy() to achieve zero copy by unmapping src_addr to the
physical page and establishing the mapping from dst_addr to the physical
page.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 fs/userfaultfd.c                 |   9 +-
 include/linux/userfaultfd_k.h    |   3 -
 include/linux/userswap.h         |  28 +++++++
 include/uapi/linux/userfaultfd.h |   1 +
 mm/Makefile                      |   1 +
 mm/mmap.c                        |   1 +
 mm/userfaultfd.c                 |  42 ++++------
 mm/userswap.c                    | 140 +++++++++++++++++++++++++++++++
 8 files changed, 191 insertions(+), 34 deletions(-)
 create mode 100644 include/linux/userswap.h
 create mode 100644 mm/userswap.c

diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index 7d0022f82844..ef51ed87ef38 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -27,13 +27,11 @@
 #include <linux/ioctl.h>
 #include <linux/security.h>
 #include <linux/hugetlb.h>
+#include <linux/userswap.h>
 
 int sysctl_unprivileged_userfaultfd __read_mostly = 1;
 
 static struct kmem_cache *userfaultfd_ctx_cachep __read_mostly;
-#ifdef CONFIG_USERSWAP
-int enable_userswap;
-#endif
 
 /*
  * Start with fault_pending_wqh and fault_wqh so they're more likely
@@ -1717,7 +1715,10 @@ static int userfaultfd_copy(struct userfaultfd_ctx *ctx,
 	ret = -EINVAL;
 	if (uffdio_copy.src + uffdio_copy.len <= uffdio_copy.src)
 		goto out;
-	if (uffdio_copy.mode & ~(UFFDIO_COPY_MODE_DONTWAKE|UFFDIO_COPY_MODE_WP))
+	if (uffdio_copy.mode & ~(UFFDIO_COPY_MODE_DONTWAKE |
+				 UFFDIO_COPY_MODE_WP |
+				 IS_ENABLED(CONFIG_USERSWAP) ?
+				 UFFDIO_COPY_MODE_DIRECT_MAP : 0))
 		goto out;
 	if (mmget_not_zero(ctx->mm)) {
 		ret = mcopy_atomic(ctx->mm, uffdio_copy.dst, uffdio_copy.src,
diff --git a/include/linux/userfaultfd_k.h b/include/linux/userfaultfd_k.h
index e1cacab86bde..e91f31a4c830 100644
--- a/include/linux/userfaultfd_k.h
+++ b/include/linux/userfaultfd_k.h
@@ -31,9 +31,6 @@
 #define UFFD_FLAGS_SET (EFD_SHARED_FCNTL_FLAGS)
 
 extern int sysctl_unprivileged_userfaultfd;
-#ifdef CONFIG_USERSWAP
-extern int enable_userswap;
-#endif
 
 extern vm_fault_t handle_userfault(struct vm_fault *vmf, unsigned long reason);
 
diff --git a/include/linux/userswap.h b/include/linux/userswap.h
new file mode 100644
index 000000000000..fe2c868851fb
--- /dev/null
+++ b/include/linux/userswap.h
@@ -0,0 +1,28 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (C) Huawei Technologies Co., Ltd. 2023. All rights reserved.
+ */
+
+#ifndef _LINUX_USERSWAP_H
+#define _LINUX_USERSWAP_H
+
+#ifdef CONFIG_USERSWAP
+
+extern int enable_userswap;
+
+int mfill_atomic_pte_nocopy(struct mm_struct *dst_mm,
+			    pmd_t *dst_pmd,
+			    struct vm_area_struct *dst_vma,
+			    unsigned long dst_addr,
+			    unsigned long src_addr);
+
+static inline bool uswap_check_copy_mode(struct vm_area_struct *vma, __u64 mode)
+{
+	if (!(vma->vm_flags & VM_USWAP) && (mode & UFFDIO_COPY_MODE_DIRECT_MAP))
+		return false;
+	return true;
+}
+
+#endif /* CONFIG_USERSWAP */
+
+#endif /* _LINUX_USERSWAP_H */
diff --git a/include/uapi/linux/userfaultfd.h b/include/uapi/linux/userfaultfd.h
index ada058f8b94b..4de57e12cdff 100644
--- a/include/uapi/linux/userfaultfd.h
+++ b/include/uapi/linux/userfaultfd.h
@@ -217,6 +217,7 @@ struct uffdio_copy {
 	 * according to the uffdio_register.ioctls.
 	 */
 #define UFFDIO_COPY_MODE_WP			((__u64)1<<1)
+#define UFFDIO_COPY_MODE_DIRECT_MAP		((__u64)1<<10)
 	__u64 mode;
 
 	/*
diff --git a/mm/Makefile b/mm/Makefile
index aad7866abe8c..0f6660ca62fa 100644
--- a/mm/Makefile
+++ b/mm/Makefile
@@ -113,6 +113,7 @@ obj-$(CONFIG_MEMORY_BALLOON) += balloon_compaction.o
 obj-$(CONFIG_PAGE_EXTENSION) += page_ext.o
 obj-$(CONFIG_CMA_DEBUGFS) += cma_debug.o
 obj-$(CONFIG_USERFAULTFD) += userfaultfd.o
+obj-$(CONFIG_USERSWAP) += userswap.o
 obj-$(CONFIG_IDLE_PAGE_TRACKING) += page_idle.o
 obj-$(CONFIG_FRAME_VECTOR) += frame_vector.o
 obj-$(CONFIG_DEBUG_PAGE_REF) += debug_page_ref.o
diff --git a/mm/mmap.c b/mm/mmap.c
index 9b6fcf8c2f1d..a5867d039153 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -49,6 +49,7 @@
 #include <linux/sched/mm.h>
 #include <linux/swapops.h>
 #include <linux/share_pool.h>
+#include <linux/userswap.h>
 
 #include <linux/uaccess.h>
 #include <asm/cacheflush.h>
diff --git a/mm/userfaultfd.c b/mm/userfaultfd.c
index 15c46208a2ac..b66abbba13ef 100644
--- a/mm/userfaultfd.c
+++ b/mm/userfaultfd.c
@@ -15,6 +15,7 @@
 #include <linux/mmu_notifier.h>
 #include <linux/hugetlb.h>
 #include <linux/shmem_fs.h>
+#include <linux/userswap.h>
 #include <asm/tlbflush.h>
 #include "internal.h"
 
@@ -90,10 +91,6 @@ static int mcopy_atomic_pte(struct mm_struct *dst_mm,
 		*pagep = NULL;
 	}
 
-#ifdef CONFIG_USERSWAP
-	if (dst_vma->vm_flags & VM_USWAP)
-		ClearPageDirty(page);
-#endif
 	/*
 	 * The memory barrier inside __SetPageUptodate makes sure that
 	 * preceding stores to the page contents become visible before
@@ -112,10 +109,6 @@ static int mcopy_atomic_pte(struct mm_struct *dst_mm,
 		else
 			_dst_pte = pte_mkwrite(_dst_pte);
 	}
-#ifdef CONFIG_USERSWAP
-	if (dst_vma->vm_flags & VM_USWAP)
-		_dst_pte = pte_mkclean(_dst_pte);
-#endif
 
 	dst_pte = pte_offset_map_lock(dst_mm, dst_pmd, dst_addr, &ptl);
 	if (dst_vma->vm_file) {
@@ -128,26 +121,9 @@ static int mcopy_atomic_pte(struct mm_struct *dst_mm,
 			goto out_release_uncharge_unlock;
 	}
 
-#ifdef CONFIG_USERSWAP
-	if (!(dst_vma->vm_flags & VM_USWAP)) {
-		ret = -EEXIST;
-		if (!pte_none(*dst_pte))
-			goto out_release_uncharge_unlock;
-	} else {
-		/*
-		 * The userspace may swap in a large area. Part of the area is
-		 * not swapped out. Skip those pages.
-		 */
-		ret = 0;
-		if (swp_type(pte_to_swp_entry(*dst_pte)) != SWP_USERSWAP_ENTRY ||
-		    pte_present(*dst_pte))
-			goto out_release_uncharge_unlock;
-	}
-#else
 	ret = -EEXIST;
 	if (!pte_none(*dst_pte))
 		goto out_release_uncharge_unlock;
-#endif
 
 	inc_mm_counter(dst_mm, MM_ANONPAGES);
 	reliable_page_counter(page, dst_mm, 1);
@@ -535,6 +511,10 @@ static __always_inline ssize_t __mcopy_atomic(struct mm_struct *dst_mm,
 		goto out_unlock;
 
 	err = -EINVAL;
+#ifdef CONFIG_USERSWAP
+	if (!uswap_check_copy_mode(dst_vma, mode))
+		goto out_unlock;
+#endif
 	/*
 	 * shmem_zero_setup is invoked in mmap for MAP_ANONYMOUS|MAP_SHARED but
 	 * it will overwrite vm_ops, so vma_is_anonymous must return false.
@@ -605,8 +585,16 @@ static __always_inline ssize_t __mcopy_atomic(struct mm_struct *dst_mm,
 		BUG_ON(pmd_none(*dst_pmd));
 		BUG_ON(pmd_trans_huge(*dst_pmd));
 
-		err = mfill_atomic_pte(dst_mm, dst_pmd, dst_vma, dst_addr,
-				       src_addr, &page, zeropage, wp_copy);
+#ifdef CONFIG_USERSWAP
+		if (dst_vma->vm_flags & VM_USWAP &&
+		    mode & UFFDIO_COPY_MODE_DIRECT_MAP)
+			err = mfill_atomic_pte_nocopy(dst_mm, dst_pmd, dst_vma,
+						      dst_addr, src_addr);
+		else
+#endif
+			err = mfill_atomic_pte(dst_mm, dst_pmd, dst_vma,
+					       dst_addr, src_addr, &page,
+					       zeropage, wp_copy);
 		cond_resched();
 
 		if (unlikely(err == -ENOENT)) {
diff --git a/mm/userswap.c b/mm/userswap.c
new file mode 100644
index 000000000000..fe33fda975d1
--- /dev/null
+++ b/mm/userswap.c
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) Huawei Technologies Co., Ltd. 2023. All rights reserved.
+ *
+ * userswap core file include swap-in and swap-out core function
+ */
+
+#include <linux/swap.h>
+#include <linux/swapops.h>
+#include <linux/rmap.h>
+#include <linux/mmu_notifier.h>
+
+#include "internal.h"
+
+int enable_userswap;
+
+static void uswap_unmap_anon_page(struct mm_struct *mm,
+				  struct vm_area_struct *vma,
+				  unsigned long addr, struct page *page,
+				  pmd_t *pmd, pte_t *old_pte)
+{
+	struct mmu_notifier_range range;
+	spinlock_t *ptl;
+	pte_t *pte;
+
+	mmu_notifier_range_init(&range, MMU_NOTIFY_UNMAP, 0, vma,
+				vma->vm_mm, addr, addr + PAGE_SIZE);
+	mmu_notifier_invalidate_range_start(&range);
+	pte = pte_offset_map_lock(mm, pmd, addr, &ptl);
+	if (pte_none(*pte))
+		goto out_release_unlock;
+	flush_cache_page(vma, addr, pte_pfn(*pte));
+	*old_pte = ptep_clear_flush(vma, addr, pte);
+
+	dec_mm_counter(mm, MM_ANONPAGES);
+	reliable_page_counter(page, mm, -1);
+	page_remove_rmap(page, false);
+
+out_release_unlock:
+	pte_unmap_unlock(pte, ptl);
+	mmu_notifier_invalidate_range_end(&range);
+	page->mapping = NULL;
+}
+
+static void uswap_map_anon_page(struct mm_struct *mm,
+				struct vm_area_struct *vma,
+				unsigned long addr,
+				struct page *page,
+				pmd_t *pmd,
+				pte_t old_pte)
+{
+	spinlock_t *ptl;
+	pte_t *pte;
+
+	pte = pte_offset_map_lock(mm, pmd, addr, &ptl);
+	flush_cache_page(vma, addr, pte_pfn(*pte));
+	set_pte_at(mm, addr, pte, old_pte);
+	inc_mm_counter(mm, MM_ANONPAGES);
+	reliable_page_counter(page, mm, 1);
+	page_add_new_anon_rmap(page, vma, addr, false);
+	pte_unmap_unlock(pte, ptl);
+}
+
+int mfill_atomic_pte_nocopy(struct mm_struct *mm,
+			    pmd_t *dst_pmd,
+			    struct vm_area_struct *dst_vma,
+			    unsigned long dst_addr,
+			    unsigned long src_addr)
+{
+	struct vm_area_struct *src_vma;
+	pte_t dst_pte, *pte, src_pte;
+	pmd_t *src_pmd;
+	spinlock_t *ptl;
+	int ret = 0;
+	struct page *page;
+
+	src_vma = find_vma(mm, src_addr);
+	if (!src_vma || src_addr < src_vma->vm_start)
+		return -ENOENT;
+
+	if (src_vma->vm_flags & VM_LOCKED)
+		return -EINVAL;
+
+	page = follow_page(src_vma, src_addr, FOLL_GET | FOLL_MIGRATION);
+	if (!page)
+		return -ENODEV;
+
+	src_pmd = mm_find_pmd(mm, src_addr);
+	if (!src_pmd) {
+		ret = -ENXIO;
+		goto out_put_page;
+	}
+	uswap_unmap_anon_page(mm, src_vma, src_addr, page, src_pmd, &src_pte);
+
+	if (dst_vma->vm_flags & VM_USWAP)
+		ClearPageDirty(page);
+	/*
+	 * The memory barrier inside __SetPageUptodate makes sure that
+	 * preceding stores to the page contents become visible before
+	 * the set_pte_at() write.
+	 */
+	__SetPageUptodate(page);
+
+	dst_pte = mk_pte(page, dst_vma->vm_page_prot);
+	if (dst_vma->vm_flags & VM_WRITE)
+		dst_pte = pte_mkwrite(pte_mkdirty(dst_pte));
+	if (dst_vma->vm_flags & VM_USWAP)
+		dst_pte = pte_mkclean(dst_pte);
+
+	pte = pte_offset_map_lock(mm, dst_pmd, dst_addr, &ptl);
+
+	/*
+	 * The userspace may swap in a large area. Part of the area is not
+	 * swapped out. If concurrent execution, PTE may be present. Skip those
+	 * pages (pte_present).
+	 * No other scenes should be handled except first pagefault (pte_none)
+	 * and after userswap out (SWP_USERSWAP_ENTRY).
+	 */
+	if (pte_present(*pte) || (!pte_none(*pte) &&
+	    swp_type(pte_to_swp_entry(*pte)) != SWP_USERSWAP_ENTRY)) {
+		pte_unmap_unlock(pte, ptl);
+		uswap_map_anon_page(mm, src_vma, src_addr, page, src_pmd,
+				    src_pte);
+		ret = -EEXIST;
+		goto out_put_page;
+	}
+
+	inc_mm_counter(mm, MM_ANONPAGES);
+	reliable_page_counter(page, mm, 1);
+	page_add_new_anon_rmap(page, dst_vma, dst_addr, false);
+	set_pte_at(mm, dst_addr, pte, dst_pte);
+
+	/* No need to invalidate - it was non-present before */
+	update_mmu_cache(dst_vma, dst_addr, pte);
+	pte_unmap_unlock(pte, ptl);
+
+out_put_page:
+	put_page(page);
+	return ret;
+}
-- 
2.17.1

